lib/scholar/neighbors/k_nearest_neighbors.ex

```diff
@@ -0,0 +1,408 @@
+defmodule Scholar.Neighbors.KNearestNeighbors do
+  @moduledoc """
+  The K-Nearest Neighbors.
+  """
+  import Nx.Defn
+  require Nx
+
+  @derive {Nx.Container,
+           keep: [:default_num_neighbors, :weights, :num_classes, :p, :task, :metric],
+           containers: [:data, :labels]}
+  defstruct [:data, :labels, :default_num_neighbors, :weights, :num_classes, :p, :task, :metric]
+
+  opts = [
+    num_neighbors: [
+      type: :pos_integer,
+      default: 5,
+      doc: "The number of neighbors to use by default for `k_neighbors` queries"
+    ],
+    num_classes: [
+      type: :pos_integer,
+      required: true,
+      doc: "Number of classes in provided labels"
+    ],
+    weights: [
+      type: {:in, [:uniform, :distance]},
+      default: :uniform,
+      doc: """
+      Weight function used in prediction. Possible values:
+
+        * `:uniform` - uniform weights. All points in each neighborhood are weighted equally.
+
+        * `:distance` - weight points by the inverse of their distance. in this case, closer neighbors of
+          a query point will have a greater influence than neighbors which are further away.
+      """
+    ],
+    metric: [
+      type: {:custom, Scholar.Options, :metric, []},
+      default: {:minkowski, 2},
+      doc: ~S"""
+      Name of the metric. Possible values:
+      * `{:minkowski, p}` - Minkowski metric. By changing value of `p` parameter we can set Euclidean, Chebyshev, Manhattan or any arbitrary $L_p$ metric.
```

Comment:
```suggestion
      Name of the metric. Possible values:

      * `{:minkowski, p}` - Minkowski metric. By changing value of `p` parameter (an integer)
        we can set Manhattan (1), Euclidean (2), Chebyshev, or any arbitrary $L_p$ metric.
```

However, isn't  "Chebyshev" infinity? How would we set it in this case?

Reply:
Yeah, we should be able to pass `:inf` as `:p`. In the `Scholar.Metrics.Distance` there is currently an error that computes Chebyshev for p = 0 which is wrong. I will fix this and then change it here as well.

Reply:
And p is actually not an integer in general, just a positive float. In sklearn it is indeed an integer, but I really don't know why.

Reply:
Ah, got it. So let’s make it a float in the validations and support `:infinity`.